CLOUDSTACK-9368: Fix for Support configurable NFS version for Secondary Storage mounts

nvazquez opened the pull request:
## Description
JIRA TICKET: https://issues.apache.org/jira/browse/CLOUDSTACK-9368
This pull request address a problem introduced in #1361 in which NFS version couldn't be changed after hosts resources were configured on startup (for hosts using `VmwareResource`), and as host parameters didn't include `nfs.version` key, it was set `null`.

## Proposed solution
In this proposed solution `nfsVersion` would be passed in `NfsTO` through `CopyCommand` to `VmwareResource`, who will check if NFS version is still configured or not. If not, it will use the one sent in the command and will set it to its storage processor and storage handler. After those setups, it will proceed executing command.

wido commented:
A version is a number, eg 3 or 4. Shouldn't the version be passed as a integer through the code instead of a String?

nvazquez commented:
You're right, I just wanted to avoid parsing but I can change it to an Integer

wido commented:
@nvazquez If you know it's a number, pass it as an int. That way you will never have a garbage string ending up somewhere. We have types for a reason :)

nvazquez commented:
Cool, I changed it and pushed again

wido commented:
Thanks! I'm not familiar with the VMWare code in CloudStack, so I won't give a LGTM. But passing a version as a Int seems a lot better.

nvazquez commented:
Thanks @wido!

rafaelweingartner commented:
@nvazquez , what about extracting that “private Integer nfsVersion” to a common superclass to all of those command class? I have created a hierarchy to help you.  The classes with an * are the ones that you did not touch directly, they are there just to build the hierarchy. I found a little odd that the “ListTemplateCommand” received those values too, are they being used there?

Maybe you cannot write a single super class to attend all of those cases, but you could easily write on to be the super class for (CopyVolumeCommand, SecStorageSetupCommand, GetStorageStatsCommand, SnapshotCommand, BackupSnapshotCommand , CreatePrivateTemplateFromSnapshotCommand,  CreatePrivateTemplateFromVolumeCommand, CreateVolumeFromSnapshotCommand)

Here is the hierachy:
- TemplateOrVolumePostUploadCommand
- Command*
	-- StorageCommand*
		--- ListTemplateCommand
	-- SsCommand*
		--- AbstractDownloadCommand*
			---- PrimaryStorageDownloadCommand
	-- CopyVolumeCommand
	-- SecStorageSetupCommand
	-- GetStorageStatsCommand
-- SnapshotCommand
--- BackupSnapshotCommand 
--- CreatePrivateTemplateFromSnapshotCommand
--- CreatePrivateTemplateFromVolumeCommand
--- CreateVolumeFromSnapshotCommand

I missing the test cases for “checkStorageProcessorAndHandlerNfsVersionAttribute”, “examineStorageSubSystemCommandNfsVersion”, “setCurrentNfsVersionInProcessorAndHandler” , “getStorageNfsVersionFromNfsTO”  and “reconfigureNfsVersion” methods. They are very good coded, clean, short, well documented. To improve them further the writing of test cases is very welcome.

nvazquez commented:
Thanks @rafaelweingartner, I like the idea, I think the best way would be leave hierarchy like this:

````
- Command
    -- NEW CLASS
        --- StorageCommand
        --- SsCommand
        --- CopyVolumeCommand
        --- SecStorageSetupCommand
        --- GetStorageStatsCommand

- TemplateOrVolumePostUploadCommand
````

This way it can be avoided `nfsVersion` parameter in every class in Command hierarchy, because they won't need it, and just leave `TemplateOrVolumePostUploadCommand` separate in this hierarchy. Would this be ok?

rafaelweingartner commented:
@nvazquez I believe so.
This way we can reduce a little bit of code.

nvazquez commented:
@rafaelweingartner I added test cases and the new hierarchy

rafaelweingartner commented on core/src/com/cloud/agent/api/storage/StorageNfsVersionCommand.java:
```diff
@@ -0,0 +1,44 @@
+
+public abstract class StorageNfsVersionCommand extends Command {
+
+    protected StorageNfsVersionCommand(){
```
are you using this constructor?

nvazquez replied:
Actually, I kept it for the hierarchy. Classes extending <code>StorageNfsVersionCommand</code> will not compile if I remove it, in their constructors the error will be:
````
Implicit super constructor StorageNfsVersionCommand() is undefined. Must explicitly invoke another constructor
````
Do you agree about keeping it?

rafaelweingartner replied:
sure I do, I had forgotten that in the child class they were invoking the super (). I am sorry for that.

nvazquez replied:
No problem ;)

rafaelweingartner commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -117,4 +154,79 @@ public void testStartVm3dgpuEnabled() throws Exception{
+    // ---------------------------------------------------------------------------------------------------
+
+    @Test
+    public void testStorageSetNfsVersion(){
```
I did not quite understand this test.
It feels like two tests in one, is it right?
Also, I believe the comments could be removed. Maybe a better name for the method would help, what do you think?

cristofolini replied:
It does seem this test would work out better if it were to be split into two. I'd recommend just testing out the two possibilities of execution for the `checkStorageProcessorAndHandlerNfsVersionAttribute` method, one test for the possibility that it will just return when the version is already set, and another one where it should set the version if it isn't already. As for naming, I'd recommend `checkStorageProcessorAndHandlerNfsVersionAttributeVersionNotSet` and `checkStorageProcessorAndHandlerNfsVersionAttributeVersionSet` respectively, but that's up to you, @nvazquez. :)

Separate tests are better in situations like this since they allow us to determine in what case(s) a method is failing to execute properly a bit more precisely.

nvazquez replied:
@rafaelweingartner @cristofolini 
The idea was testing that once a version is set it cannot be changed later, but as you mentioned below I tryied testing the whole flow, instead of the method. I liked @cristofolini recommendation, I'll work on it

rafaelweingartner commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -117,4 +154,79 @@ public void testStartVm3dgpuEnabled() throws Exception{
+    }
+
+    @Test
+    public void testStorageNfsVersionNotPresent(){
```
I believe this method could be simpler. 
If it wants to test what is coded now. It would only need check if "_storageNfsVersion == null", then, Mockito.verify(mock, times(0)).examineStorageSubSystemCommandNfsVersion

Also for the other case you can use Mockito.doNothing, when calling "examineStorageSubSystemCommandNfsVersion". This is an integration test, you do not need to test the whole execution flow, you only need to test if the method (the one being tested) is calling the methods it should.

That means, to test the "checkStorageProcessorAndHandlerNfsVersionAttribute", you only need to worry about the "examineStorageSubSystemCommandNfsVersion" method.

nvazquez replied:
Nice, I'll work on this

nvazquez commented:
@rafaelweingartner @cristofolini I refactored unit tests based on your comments, I also removed <code>testStorageNfsVersionNotPresent</code> as it is covered by tests <code>checkStorageProcessorAndHandlerNfsVersionAttributeVersionSet</code> and <code>checkStorageProcessorAndHandlerNfsVersionAttributeVersionNotSet</code>

rafaelweingartner commented on plugins/hypervisors/vmware/test/com/cloud/hypervisor/vmware/resource/VmwareResourceTest.java:
```diff
@@ -117,4 +153,60 @@ public void testStartVm3dgpuEnabled() throws Exception{
+    // ---------------------------------------------------------------------------------------------------
+
+    @Test
+    public void checkStorageProcessorAndHandlerNfsVersionAttributeVersionNotSet(){
```
@nvazquez if the version is not set, should not we check if the method "examineStorageSubSystemCommandNfsVersion" was called?

nvazquez replied:
Sounds good

nvazquez replied:
I added it

rafaelweingartner commented:
@nvazquez, that is great.
As always your PRs are perfect, great job;)
LGTM from me here.

rafaelweingartner commented:
Just an update, 
I forgot to ask, can you squash the commits?
I think this one it is better to have all of the changes into a single commit.

nvazquez commented:
@rafaelweingartner I squashed them, thanks for your comments and reviews!

serg38 commented:
All checks passed. @rhtyd, @wido or @GabrielBrascher   can you review and give second OK.

serg38 commented:
@koushik-das  Will you be able to review this PR? It has been waiting for 2d LGTM for a while now.

koushik-das commented on engine/storage/image/src/org/apache/cloudstack/storage/image/TemplateServiceImpl.java:
```diff
@@ -638,7 +638,9 @@ protected Void createTemplateAsyncCallBack(AsyncCallbackDispatcher<TemplateServi
     private Map<String, TemplateProp> listTemplate(DataStore ssStore) {
-        ListTemplateCommand cmd = new ListTemplateCommand(ssStore.getTO(), imageStoreDetailsUtil.getNfsVersion(ssStore.getId()));
+        String nfsVersionParam = imageStoreDetailsUtil.getNfsVersion(ssStore.getId());
+        Integer nfsVersion = (nfsVersionParam != null ? Integer.valueOf(nfsVersionParam) : null);
```
Please move the nfs version conversion from string to int to a separate method, saw the same pattern at multiple places.

koushik-das commented on engine/storage/src/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java:
```diff
@@ -79,6 +80,25 @@
+    @Inject
+    ImageStoreDetailsDao _imageStoreDetailsDao;
+
+    private static final String NFS_VERSION_DETAILS_KEY = "nfs.version";
```
I know this was added as part of a previous commit, but not sure if this the right place to put nfs version related details. What if the image store type is not NFS but SWIFT/S3/CIFS? I think there needs to be a better abstraction of the properties based on store type.

nvazquez replied:
@koushik-das sorry for late reply. A reason for defining nfs details in `BaseImageStoreDriverImpl` was that as `CloudStackImageStoreDriverImpl` and `SimulatorImageStoreDriverImpl` both extend it and they both return a NfsTO instance on `getStoreTO` we would need that key at least for the 2 mentioned classes. What would you suggest?

koushik-das replied:
@nvazquez One way would be create an intermediate class for NFS based stores and derive CloudStackImageStoreDriverImpl and SimulatorImageStoreDriverImpl from that.

koushik-das commented on services/secondary-storage/server/src/org/apache/cloudstack/storage/template/DownloadManagerImpl.java:
```diff
@@ -984,7 +984,8 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
             s_logger.info("DownloadManager: starting additional services since we are inside system vm");
-            _nfsVersion = (String)params.get("nfsVersion");
+            String nfsVersionParam = (String)params.get("nfsVersion");
+            _nfsVersion = (nfsVersionParam != null ? Integer.valueOf(nfsVersionParam) : null);
```
Saw this pattern getting repeated in multiple places, if possible move it to a method.

koushik-das commented:
Looks like nfs version is only supported for VMware. If XS or KVM clusters are added to the zone (having nfs version set on image store) will there be any side effects?
What if this property is set on a image store which is not NFS?

@sateesh-chodapuneedi can you review the VMware code?

nvazquez commented:
@koushik-das I only use Vmware hypervisor so I couldn't test it for Xen or KVM. However, I reviewed code and found `copySnapshotToTemplateFromNfsToNfsXenserver` method in `NfsSecondaryStorageResource` which will attemp mount using nfs version. I'm not very familiar to Xen code in CloudStack but if it uses NfsSecondaryStorage for mounting I think I can succeed, but it should be tested or improved to support those hypervisors

swill commented:
I will run this through my KVM CI right now...

swill commented:
I have tried to CI this PR a couple times.  I get the following during Deploy DC.
```
==== Deploy DC Started ====
Exception Occurred :['Traceback (most recent call last):\n', '  File "/data/git/cs1/cloudstack/tools/marvin/marvin/deployDataCenter.py", line 142, in addHosts\n    ret = self.__apiClient.addHost(hostcmd)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 1530, in addHost\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: addhost failed, due to: errorCode: 530, errorText:Unable to add the host\n']
Exception Occurred :['Traceback (most recent call last):\n', '  File "/data/git/cs1/cloudstack/tools/marvin/marvin/deployDataCenter.py", line 142, in addHosts\n    ret = self.__apiClient.addHost(hostcmd)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 1530, in addHost\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: addhost failed, due to: errorCode: 530, errorText:Guid is not updated for cluster with specified cluster id; need to wait for hosts in this cluster to come up\n']
```

I can't find anything in the `jetty.log` to show why it is failing.  I am a bit confused honestly...

serg38 commented:
This doesn't seem to be related to this PR.Can you post managment-server.log extract around the failure time?

swill commented:
I am running into this on a couple PRs I need to figure out what is going on...

GabrielBrascher commented on plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -519,6 +523,64 @@ public Answer executeRequest(Command cmd) {
+    /**
+     * Get storage NFS version from NfsTO
+     * @param nfsTO nfsTO
+     * @return true if NFS version if found and not null, false in other case
```
@nvazquez just a small typo, change "_true if NFS version **if** found_" to "_true if NFS version **was** found_".
Good job with the code documentation, thanks.

GabrielBrascher commented on plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -303,6 +305,7 @@
     protected String _password;
     protected String _guid;
     protected String _vCenterAddress;
+    protected Integer _storageNfsVersion;
```
There has been a discussion about slowly remove the underscore at the beginning of variables. Could you please remove it? Thanks.

GabrielBrascher commented on plugins/hypervisors/vmware/src/com/cloud/storage/resource/VmwareStorageSubsystemCommandHandler.java:
```diff
@@ -61,11 +61,28 @@ public void setStorageManager(VmwareStorageManager storageManager) {
+    /**
+     * Reconfigure NFS version for storage operations
+     * @param nfsVersion NFS version to set
+     * @return true if NFS version coulb de configured, false in other case
```
Should it be "_true if NFS version coul**d b**e configured, false in other case_"? Thanks for the javadoc, good work.

GabrielBrascher commented:
@nvazquez sorry for the late review,

I could point just 2 typos on Javadoc and some variables with the underscore (would be nice to change those variables names, but they are nothing serious).

Great work, I can't see anything wrong, the code LGTM.

**Note:** I am just given my feedback, I don't want to run over the @koushik-das review, it would be nice to check his comment about _NFS_VERSION_DETAILS_KEY_ at the "BaseImageStoreDriverImpl" class. Thanks.

nvazquez commented:
@GabrielBrascher thanks for your review! I pushed changes fixing issues you've found. 
@koushik-das I've replied your comment, sorry for late reply

koushik-das commented:
@nvazquez I am not too familiar with the VMware specific code, rest of the code looks ok.

Also a comment on #1361 when this feature was initially added. NFS version is stored in image_store_details table and needs to be manually added using some sql script. A better way would be to use ConfigKey to expose the configuration. There will be some work involved to include image store as a scope for specifying configuration. May be you can fix it in subsequent PR.

swill commented:
I am having a hard time testing this because I can't seem to get the templates to become ready using this PR.  My CI ran all night last night and this morning it was still waiting for the templates to be ready.  I just did another run and I am stuck at the same place.

```
http://localhost:8096/client/api?command=listTemplates&templatefilter=all&response=json
Tue May 17 17:17:16 2016 At least template 'tiny linux kvm' is not Ready
```

I am not seeing any errors in the `jetty.log`, so I am not sure what is going on here...

nvazquez commented:
@koushik-das I pushed new changes including an intermediate class as you suggested
@swill hopefully CI passes now

DaanHoogland commented on engine/storage/src/org/apache/cloudstack/storage/image/NfsImageStoreDriverImpl.java:
```diff
@@ -0,0 +1,30 @@
+package org.apache.cloudstack.storage.image;
```
license missing

nvazquez replied:
Done, thanks!

koushik-das commented:
@nvazquez Thanks for the fix. As @DaanHoogland mentioned please add license header to the newly added file.

swill commented:
### CI RESULTS

```
Tests Run: 83
  Skipped: 0
   Failed: 0
   Errors: 2
 Duration: 8h 53m 51s
```

**Summary of the problem(s):**
```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcRemoteAccessVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 293, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_WU9CS8/results.txt
```

```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcSite2SiteVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 472, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_WU9CS8/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_19_2016_22_26_12_OT4DJY:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/DeployDataCenter__May_19_2016_22_26_12_OT4DJY/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/DeployDataCenter__May_19_2016_22_26_12_OT4DJY/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/DeployDataCenter__May_19_2016_22_26_12_OT4DJY/runinfo.txt)

**`/tmp/MarvinLogs/test_network_WU9CS8:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_network_WU9CS8/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_network_WU9CS8/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_network_WU9CS8/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_U00WPJ:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_vpc_routers_U00WPJ/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_vpc_routers_U00WPJ/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1518/tmp/MarvinLogs/test_vpc_routers_U00WPJ/runinfo.txt)


Uploads will be available until `2016-07-20 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

swill commented:
This is a passing CI now.  @koushik-das do you have anything outstanding with this PR?

@nvazquez can you close and reopen or do a force push to kick off a Jenkins run to see if we can get this green?  Thanks for the work on this to get it ready...

nvazquez commented:
Thanks @swill @koushik-das! Sorry I was missing a license header, I pushed it now.

koushik-das commented:
@swill nothing outstanding on this

swill commented:
For some reason Travis is unhappy.  Would you mind trying again.  Sorry for the inconvenience...

serg38 commented:
@swill Travis passed. Looks like it is ready to merge.
